node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,735 @@
+
+			if features == "" {
+				features = "ibc:"
+			} else {
+				features += "|"
+			}
+			features += ce.chainID.String()
+
+			p2p.DefaultRegistry.SetNetworkStats(ce.chainID, &gossipv1.Heartbeat_Network{ContractAddress: w.contractAddress})
+		}
+
+		Features = features
+	}
+
+	w.logger.Info("creating watcher",
+		zap.String("wsUrl", w.wsUrl),
+		zap.String("lcdUrl", w.lcdUrl),
+		zap.String("contract", w.contractAddress),
+		zap.String("features", Features),
+	)
+
+	c, _, err := websocket.Dial(ctx, w.wsUrl, nil)
+	if err != nil {
+		connectionErrors.WithLabelValues("websocket_dial_error").Inc()
+		return fmt.Errorf("failed to establish tendermint websocket connection: %w", err)
+	}
+	defer c.Close(websocket.StatusNormalClosure, "")
+
+	c.SetReadLimit(cosmwasm.ReadLimitSize)
+
+	// Subscribe to smart contract transactions.
+	params := [...]string{fmt.Sprintf("tm.event='Tx' AND wasm._contract_address='%s'", w.contractAddress)}
+	command := &clientRequest{
+		JSONRPC: "2.0",
+		Method:  "subscribe",
+		Params:  params,
+		ID:      1,
+	}
+	err = wsjson.Write(ctx, c, command)
+	if err != nil {
+		connectionErrors.WithLabelValues("websocket_subscription_error").Inc()
+		return fmt.Errorf("failed to subscribe to events: %w", err)
+	}
+
+	// Wait for the success response.
+	_, subResp, err := c.Read(ctx)
+	if err != nil {
+		connectionErrors.WithLabelValues("websocket_subscription_error").Inc()
+		return fmt.Errorf("failed to receive response to subscribe request: %w", err)
+	}
+	if strings.Contains(string(subResp), "error") {
+		connectionErrors.WithLabelValues("websocket_subscription_error").Inc()
+		return fmt.Errorf("failed to subscribe to events, response: %s", string(subResp))
+	}
+
+	// Start a routine to listen for messages from the contract.
+	common.RunWithScissors(ctx, errC, "ibc_data_pump", func(ctx context.Context) error {
+		return w.handleEvents(ctx, c, errC)
+	})
+
+	// Start a routine to periodically query the wormchain block height.
+	common.RunWithScissors(ctx, errC, "ibc_block_height", func(ctx context.Context) error {
+		return w.handleQueryBlockHeight(ctx, c, errC)
+	})
+
+	// Start a routine for each chain to listen for observation requests.
+	for _, ce := range w.chainMap {
+		common.RunWithScissors(ctx, errC, "ibc_objs_req", func(ctx context.Context) error {
+			return w.handleObservationRequests(ctx, errC, ce)
+		})
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
+}
+
+// handleEvents reads messages from the IBC receiver contract and processes them.
+func (w *Watcher) handleEvents(ctx context.Context, c *websocket.Conn, errC chan error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			_, message, err := c.Read(ctx)
+			if err != nil {
+				w.logger.Error("failed to read socket", zap.Error(err))
+				connectionErrors.WithLabelValues("channel_read_error").Inc()
+				errC <- fmt.Errorf("failed to read socket: %w", err)
+				return nil
```

Comment:
Wouldn't you want to return an error message?  Or you don't need to because runWithScissors() will put that on the errC and you already did that here?

Reply:
Changed it.